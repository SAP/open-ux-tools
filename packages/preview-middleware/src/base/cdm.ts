import { readFileSync } from 'fs';
import { join } from 'path';
import type { TemplateConfig } from './config';
import { type FLPCdmConfig, FLPHomePageDefaults } from '../types';

/**
 * Generates a CDM by embedding the provided app tiles into the FLP homepage.
 *
 * @param apps - A list of app to be embedded.
 * @returns The generated CDM configuration
 */
export function generateCdm(apps: TemplateConfig['apps']): FLPCdmConfig {
    const cdm = JSON.parse(readFileSync(join(__dirname, '../../templates/flp/cdm.base.json'), 'utf-8')) as FLPCdmConfig;

    // add apps
    Object.keys(apps).forEach((id) => {
        const appId = apps[id].additionalInformation.split('=')[1];
        const vizId = `VIZ:${appId}`;
        const [object, action] = id.split('-');
        const { title, description, url } = apps[id];

        // add app to default catalog
        cdm.catalogs[FLPHomePageDefaults.catalogId].payload.viz.push(vizId);

        // create flp visualization
        cdm.visualizations[vizId] = {
            'vizType': 'sap.ushell.StaticAppLauncher',
            'businessApp': appId,
            'vizConfig': {
                'sap.app': {
                    title,
                    subTitle: description
                },
                'sap.flp': {
                    'target': {
                        'appId': appId,
                        'inboundId': `${object}-${action}`,
                        'parameters': [
                            {
                                'name': 'sap-ui-tech-hint',
                                'value': 'UI5'
                            }
                        ]
                    }
                }
            }
        };

        // create flp application
        cdm.applications[appId] = {
            'sap.app': {
                id: appId,
                title,
                crossNavigation: {
                    inbounds: {
                        [`${object}-${action}`]: {
                            'semanticObject': object,
                            'action': action,
                            title,
                            'subTitle': description,
                            'signature': {
                                'additionalParameters': 'allowed'
                            }
                        }
                    }
                }
            },
            'sap.ui5': {
                'componentName': appId
            },
            'sap.ui': {
                'technology': 'UI5'
            },
            'sap.platform.runtime': {
                'componentProperties': {
                    url
                }
            }
        };

        // add app to default section
        cdm.pages[FLPHomePageDefaults.pageName].payload.sections[FLPHomePageDefaults.sectionId].layout.vizOrder.push(
            appId
        );
        cdm.pages[FLPHomePageDefaults.pageName].payload.sections[FLPHomePageDefaults.sectionId].viz[appId] = {
            id: appId,
            vizId
        };
    });

    return cdm;
}
