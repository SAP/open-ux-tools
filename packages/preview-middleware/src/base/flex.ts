import type { Logger } from '@sap-ux/logger';
import type { ReaderCollection } from '@ui5/fs';
import { existsSync, mkdirSync, readdirSync, unlinkSync, writeFileSync } from 'fs';
import { join, parse } from 'path';

/**
 * Structure of a flex change.
 */
export interface FlexChange {
    [key: string]: string | object | undefined;
    changeType: string;
    reference: string;
    moduleName?: string;
    content?: {
        codeRef?: string;
        fragmentPath?: string;
    };
}

/**
 * Map of change type specific correction functions.
 */
const moduleNameContentMap: { [key: string]: (change: FlexChange) => string } = {
    codeExt: (change) => (change.content?.codeRef ?? '').replace('.js', ''),
    addXML: (change) => change.content?.fragmentPath ?? ''
};

/**
 * Sets the moduleName property of the provided change to also support old changes with newer UI5 versions.
 *
 * @param change change to be fixed
 * @param logger logger instance
 */
function tryFixChange(change: FlexChange, logger: Logger) {
    try {
        const prefix = change.reference.replace(/\./g, '/');
        change.moduleName = `${prefix}/changes/${moduleNameContentMap[change.changeType](change)}`;
    } catch (error) {
        logger.warn('Could not fix missing module name.');
    }
}

/**
 * Read changes from the file system and return them.
 *
 * @param project reference to the UI5 project
 * @param logger logger instance
 * @returns object with the file name as key and the file content as value
 */
export async function readChanges(project: ReaderCollection, logger: Logger): Promise<Record<string, FlexChange>> {
    const changes: Record<string, FlexChange> = {};
    const files = await project.byGlob('/**/changes/*.*');
    for (const file of files) {
        try {
            const change = JSON.parse(await file.getString()) as FlexChange;
            if (moduleNameContentMap[change.changeType] && !change.moduleName) {
                tryFixChange(change, logger);
            }
            changes[`sap.ui.fl.${parse(file.getName()).name}`] = change;
            logger.debug(`Read change from ${file.getPath()}`);
        } catch (error) {
            logger.warn(error.message);
        }
    }
    return changes;
}

/**
 * Writes flex changes to the user's workspace.
 *
 * @param data JS object to be stored as change
 * @param data.fileName file name that is required for a valid change
 * @param data.fileType file type that is required
 * @param webappPath path to the webapp folder
 * @param logger logger instance
 * @returns object with success flag and optional message
 */
export function writeChange(
    data: object & { fileName?: string; fileType?: string },
    webappPath: string,
    logger: Logger
): { success: boolean; message?: string } {
    const fileName = data.fileName;
    const fileType = data.fileType;
    if (fileName && fileType) {
        logger.debug(`Write change ${fileName}.${fileType}`);
        const path = join(webappPath, 'changes');
        if (!existsSync(path)) {
            mkdirSync(path);
        }
        const filePath = join(path, fileName + '.' + fileType);
        writeFileSync(filePath, JSON.stringify(data, null, 2));
        const message = `FILE_CREATED ${fileName}.${fileType}`;
        return { success: true, message };
    } else {
        return { success: false };
    }
}

/**
 * Deletes an existing flex change from the file system if it exists.
 *
 * @param data JS object of the change to be deleted
 * @param data.fileName file name that is required for a valid change
 * @param webappPath path to the webapp folder
 * @param logger logger instance
 * @returns object with success flag and optional message
 */
export function deleteChange(
    data: object & { fileName?: string },
    webappPath: string,
    logger: Logger
): { success: boolean; message?: string } {
    const fileName = data.fileName?.replace('sap.ui.fl.', '');
    if (fileName) {
        const path = join(webappPath, 'changes');
        if (existsSync(path)) {
            const files = readdirSync(path);
            const file = files.find((element) => element.includes(fileName));
            if (file) {
                logger.debug(`Write change ${file}`);
                const filePath = join(path, file);
                unlinkSync(filePath);
                return { success: true, message: `FILE_DELETED ${file}` };
            }
        }
    }
    return { success: false };
}
