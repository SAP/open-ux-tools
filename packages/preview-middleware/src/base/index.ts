export { FlpSandbox, initAdp } from './flp';
