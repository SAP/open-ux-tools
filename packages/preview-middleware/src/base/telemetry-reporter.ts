import { isInternalFeaturesSettingEnabled } from '@sap-ux/feature-toggle';
import { ClientFactory, EventName, SampleRate, initTelemetrySettings } from '@sap-ux/telemetry';
import modulePackageJson from '../../package.json';
import dotenv from 'dotenv';
import { join } from 'path';
import { enableTelemetry } from '@sap-ux-private/control-property-editor-common';

export class TelemetryReporter {
    constructor(private eventName: string, private layer?: string) {}
    public async initializeTelemetry() {
        const envFilePath = { path: join(__dirname, '..', '..', '.env') };
        dotenv.config(envFilePath);
        const internalFeature = isInternalFeaturesSettingEnabled();
        await initTelemetrySettings({
            consumerModule: modulePackageJson,
            internalFeature,
            watchTelemetrySettingStore: true,
            resourceId: process.env['OpenUxTools_ResourceId']
        });
        enableTelemetry();
    }

    public reportTelemetry(data: any) {
        const telemetryEvent = {
            eventName: this.eventName,
            properties: { ...data, layer: this.layer },
            measurements: {}
        };
        ClientFactory.getTelemetryClient().reportEvent(telemetryEvent, SampleRate.NoSampling, {
            appPath: process.cwd()
        });
    }
}
