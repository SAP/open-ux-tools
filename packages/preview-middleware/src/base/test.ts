import type { Resource } from '@ui5/fs';
import type { InternalTestConfig, TestConfig } from '../types';

const DEFAULTS: Record<string, InternalTestConfig> = {
    qunit: {
        path: '/test/unitTests.qunit.html',
        init: '/test/unitTests.qunit.js',
        pattern: '/test/**/*Test.*',
        framework: 'QUnit'
    },
    opa5: {
        path: '/test/opaTests.qunit.html',
        init: '/test/opaTests.qunit.js',
        pattern: '/test/**/*Journey.*',
        framework: 'OPA5'
    }
};

/**
 * Merge the given test configuration with the default values.
 *
 * @param config test configuration
 * @returns merged test configuration
 */
export function mergeTestConfigDefaults(config: TestConfig) {
    const key = config.framework.toLowerCase();
    const defaults = DEFAULTS[key] ?? {};
    if (config.path && !config.path.startsWith('/')) {
        config.path = `/${config.path}`;
    }
    if (config.init && !config.init.startsWith('/')) {
        config.init = `/${config.init}`;
    }
    return {
        framework: config.framework,
        path: config.path ?? defaults.path,
        init: config.init ?? defaults.init,
        pattern: config.pattern ?? defaults.pattern
    } satisfies TestConfig;
}

/**
 * Generate a list of imports from a list of resources.
 *
 * @param ns namespace of the test files
 * @param resourceList list of resources representing test files
 * @returns array of strings representing the tests
 */
export function generateImportList(ns: string, resourceList: Resource[]): string[] {
    return resourceList
        ? resourceList.map((file) => {
              const path = file.getPath().split('.');
              path.pop();
              return `${ns}${path.join('.')}`;
          })
        : [];
}
