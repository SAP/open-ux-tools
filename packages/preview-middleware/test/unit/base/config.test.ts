import type { Manifest } from '@sap-ux/project-access';
import {
    DEFAULT_INTENT,
    DEFAULT_PATH,
    createFlpTemplateConfig,
    createTestTemplateConfig,
    generatePreviewFiles,
    getFlpConfigWithDefaults,
    getPreviewPaths
} from '../../../src/base/config';
import { mergeTestConfigDefaults } from '../../../src/base/test';
import type { MiddlewareConfig } from '../../../src';
import { join } from 'path';
import { ToolsLogger } from '@sap-ux/logger';

describe('config', () => {
    const manifest = {
        'sap.app': {
            id: 'my.app'
        }
    } as Manifest;
    describe('getFlpConfigWithDefaults', () => {
        test('minimum settings', () => {
            const flpConfig = getFlpConfigWithDefaults({});
            expect(flpConfig).toMatchObject({
                path: DEFAULT_PATH,
                intent: DEFAULT_INTENT
            });
        });
        test('user configured path', () => {
            const path = '/test/flpSandbox.html';
            const intent = { object: 'myapp', action: 'myaction' };
            const flpConfig = getFlpConfigWithDefaults({ path, intent });
            expect(flpConfig).toMatchObject({ path, intent });
        });
    });

    describe('createFlpTemplateConfig', () => {
        test('minimum settings', () => {
            const flpConfig = getFlpConfigWithDefaults({});
            const templateConfig = createFlpTemplateConfig(flpConfig, manifest);
            expect(templateConfig).toMatchSnapshot();
        });

        test('minimum settings with one reuse lib', () => {
            const flpConfig = getFlpConfigWithDefaults({});
            const resources = { 'my.reuse.lib': '/custom/path/my.reuse.lib' };
            const templateConfig = createFlpTemplateConfig(flpConfig, manifest, resources);
            expect(templateConfig).toMatchSnapshot();
        });
    });

    describe('createTestTemplateConfig', () => {
        test('minimum settings', () => {
            const config = mergeTestConfigDefaults({ framework: 'OPA5' });
            const templateConfig = createTestTemplateConfig(config, manifest['sap.app'].id, 'sap_horizon');
            expect(templateConfig).toMatchObject({
                basePath: '..',
                framework: 'OPA5',
                id: manifest['sap.app'].id,
                initPath: 'opaTests.qunit.js',
                theme: 'sap_horizon'
            });
        });
    });

    describe('getPreviewPaths', () => {
        test('minimum settings', async () => {
            const paths = getPreviewPaths({});
            expect(paths).toHaveLength(1);
            expect(paths[0]).toMatchObject({
                path: `${DEFAULT_PATH}#${DEFAULT_INTENT.object}-${DEFAULT_INTENT.action}`,
                type: 'preview'
            });
        });

        test('tests included and a custom path', async () => {
            const consoleSpyError = jest.spyOn(ToolsLogger.prototype, 'error').mockImplementation(() => {});
            const consoleSpyWarning = jest.spyOn(ToolsLogger.prototype, 'warn').mockImplementation(() => {});
            const config = {
                flp: {
                    path: '/test/flpSandbox.html',
                    intent: { object: 'myapp', action: 'myaction' }
                },
                rta: {
                    layer: 'CUSTOMER_BASE',
                    editors: [
                        { path: '/local/editor.html', developerMode: false },
                        { path: '/local/developerEditor.html', developerMode: true }
                    ]
                },
                test: [{ framework: 'OPA5' }]
            } as MiddlewareConfig;
            const previews = getPreviewPaths(config);
            expect(previews).toHaveLength(4);
            expect(
                previews.find(
                    ({ path }) =>
                        path === `${config?.flp?.path}#${config?.flp?.intent?.object}-${config?.flp?.intent?.action}`
                )
            ).toBeDefined();
            expect(previews.find(({ path }) => path === config?.editors?.rta?.endpoints[0]?.path)).toBeDefined();
            expect(previews.find(({ path }) => path === config?.editors?.rta?.endpoints[1]?.path)).toBeDefined();
            expect(consoleSpyError).toHaveBeenCalledWith(
                'developerMode is ONLY supported for SAP UI5 adaptation projects.'
            );
            expect(consoleSpyWarning).toHaveBeenCalledWith('developerMode for /local/developerEditor.html disabled');
            expect(previews.find(({ path }) => path === '/test/opaTests.qunit.html')).toBeDefined();
            consoleSpyError.mockRestore();
            consoleSpyWarning.mockRestore();
        });
    });

    describe('generatePreviewFiles', () => {
        test('minimum settings', async () => {
            const basePath = join(__dirname, '../../fixtures/simple-app');
            const fs = await generatePreviewFiles(basePath, {});
            const files = fs.dump(basePath);
            const paths = Object.keys(files);
            expect(paths).toHaveLength(1);
            expect(paths[0].endsWith(DEFAULT_PATH)).toBe(true);
            expect(files[paths[0]].contents).toMatchSnapshot();
        });

        test('tests included and a custom path', async () => {
            const basePath = join(__dirname, '../../fixtures/simple-app');
            const config = {
                flp: {
                    path: '/test/flpSandbox.html',
                    intent: { object: 'myapp', action: 'myaction' }
                },
                test: [{ framework: 'OPA5' }, { framework: 'Testsuite' }]
            } satisfies MiddlewareConfig;
            const fs = await generatePreviewFiles(basePath, config);
            expect(fs.dump(basePath)).toMatchSnapshot();
        });

        test('multi-app setup e.g. in CAP', async () => {
            const basePath = join(__dirname, '../../fixtures');
            const config = {
                flp: {
                    path: '/test/flpSandbox.thml',
                    apps: [
                        {
                            local: '/simple-app',
                            target: '/apps/simple-app',
                            intent: {
                                object: 'simpleApp',
                                action: 'preview'
                            }
                        },
                        {
                            local: '/multi-app',
                            target: '/apps/other-app'
                        }
                    ]
                }
            } satisfies MiddlewareConfig;
            const fs = await generatePreviewFiles(basePath, config);
            expect(fs.dump(basePath)).toMatchSnapshot();
        });
    });
});
