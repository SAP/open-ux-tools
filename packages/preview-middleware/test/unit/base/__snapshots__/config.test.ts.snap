// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config createFlpTemplateConfig minimum settings 1`] = `
Object {
  "apps": Object {},
  "basePath": "..",
  "bootstrapPath": "../resources/sap/ushell/bootstrap/sandbox2.js",
  "init": undefined,
  "locateReuseLibsScript": undefined,
  "ui5": Object {
    "bootstrapOptions": "",
    "flex": Array [
      Object {
        "connector": "LrepConnector",
        "layers": Array [],
        "url": "/sap/bc/lrep",
      },
      Object {
        "applyConnector": "custom.connectors.WorkspaceConnector",
        "custom": true,
        "writeConnector": "custom.connectors.WorkspaceConnector",
      },
      Object {
        "connector": "LocalStorageConnector",
        "layers": Array [
          "CUSTOMER",
          "USER",
        ],
      },
    ],
    "libs": "sap.m,sap.ui.core,sap.ushell",
    "resources": Object {
      "open.ux.preview.client": "/preview/client",
    },
    "theme": "sap_horizon",
  },
}
`;

exports[`config generatePreviewFiles minimum settings 1`] = `
"<!DOCTYPE HTML>
<html lang=\\"en\\">
<!-- Copyright (c) 2015 SAP AG, All Rights Reserved -->

<head>
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <title>Local FLP Sandbox</title>

    <!-- Bootstrap the unified shell in sandbox mode for standalone usage.

         The renderer is specified in the global Unified Shell configuration object \\"sap-ushell-config\\".

         The fiori2 renderer will render the shell header allowing, for instance,
         testing of additional application setting buttons.

         The navigation target resolution service is configured in a way that the empty URL hash is
         resolved to our own application.

         This example uses relative path references for the SAPUI5 resources and test-resources;
         it might be necessary to adapt them depending on the target runtime platform.
         The sandbox platform is restricted to development or demo use cases and must NOT be used
         for productive scenarios.
    -->
    <script type=\\"text/javascript\\">
        window[\\"sap-ushell-config\\"] = {
            defaultRenderer: \\"fiori2\\",
            renderers: {
                fiori2: {
                    componentData: {
                        config: {
                            search: \\"hidden\\",
                            enableSearch: false
                        }
                    }
              }
          },
          applications: {\\"app-preview\\":{\\"title\\":\\"My Simple App\\",\\"description\\":\\"This is a very simple application.\\",\\"additionalInformation\\":\\"SAPUI5.Component=test.fe.v2.app\\",\\"applicationType\\":\\"URL\\",\\"url\\":\\"..\\",\\"applicationDependencies\\":{\\"manifest\\":true}}}
        };
    </script>

    <script src=\\"../test-resources/sap/ushell/bootstrap/sandbox.js\\" id=\\"sap-ushell-bootstrap\\"></script>
    <!-- Bootstrap the UI5 core library. 'data-sap-ui-frameOptions=\\"allow\\"'' is a NON-SECURE setting for test environments -->
    <script id=\\"sap-ui-bootstrap\\"
        src=\\"../resources/sap-ui-core.js\\"
        data-sap-ui-libs=\\"sap.m,sap.ui.core,sap.ushell,sap.f,sap.ui.comp,sap.ui.generic.app,sap.suite.ui.generic.template\\"
        data-sap-ui-async=\\"true\\"
        data-sap-ui-preload=\\"async\\"
        data-sap-ui-theme=\\"sap_horizon_dark\\"
        data-sap-ui-compatVersion=\\"edge\\"
        data-sap-ui-language=\\"en\\"
        data-sap-ui-bindingSyntax=\\"complex\\"
        data-sap-ui-flexibilityServices='[{\\"connector\\":\\"LrepConnector\\",\\"layers\\":[],\\"url\\":\\"/sap/bc/lrep\\"},{\\"applyConnector\\":\\"custom.connectors.WorkspaceConnector\\",\\"writeConnector\\":\\"custom.connectors.WorkspaceConnector\\",\\"custom\\":true},{\\"connector\\":\\"LocalStorageConnector\\",\\"layers\\":[\\"CUSTOMER\\",\\"USER\\"]}]'
        data-sap-ui-resourceroots='{\\"open.ux.preview.client\\":\\"/preview/client\\",\\"test.fe.v2.app\\":\\"..\\"}'
        data-sap-ui-frameOptions=\\"allow\\"
        data-sap-ui-xx-componentPreload=\\"off\\"
        data-sap-ui-oninit=\\"module:open/ux/preview/client/flp/init\\">
    </script>

    <script type=\\"text/javascript\\">
        (() => sap.ui.require(['open/ux/preview/client/flp/initConnectors']))()
    </script>

    
</head>

<!-- UI Content -->

<body class=\\"sapUiBody\\" id=\\"content\\">
</body>

</html>"
`;

exports[`config generatePreviewFiles multi-app setup e.g. in CAP 1`] = `
Object {
  "test/flpSandbox.thml": Object {
    "contents": "<!DOCTYPE HTML>
<html lang=\\"en\\">
<!-- Copyright (c) 2015 SAP AG, All Rights Reserved -->

<head>
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <title>Local FLP Sandbox</title>

    <!-- Bootstrap the unified shell in sandbox mode for standalone usage.

         The renderer is specified in the global Unified Shell configuration object \\"sap-ushell-config\\".

         The fiori2 renderer will render the shell header allowing, for instance,
         testing of additional application setting buttons.

         The navigation target resolution service is configured in a way that the empty URL hash is
         resolved to our own application.

         This example uses relative path references for the SAPUI5 resources and test-resources;
         it might be necessary to adapt them depending on the target runtime platform.
         The sandbox platform is restricted to development or demo use cases and must NOT be used
         for productive scenarios.
    -->
    <script type=\\"text/javascript\\">
        window[\\"sap-ushell-config\\"] = {
            defaultRenderer: \\"fiori2\\",
            renderers: {
                fiori2: {
                    componentData: {
                        config: {
                            search: \\"hidden\\",
                            enableSearch: false
                        }
                    }
              }
          },
          applications: {\\"simpleApp-preview\\":{\\"title\\":\\"My Simple App\\",\\"description\\":\\"This is a very simple application.\\",\\"additionalInformation\\":\\"SAPUI5.Component=test.fe.v2.app\\",\\"applicationType\\":\\"URL\\",\\"url\\":\\"/apps/simple-app\\",\\"applicationDependencies\\":{\\"manifest\\":true}},\\"testfev2other-preview\\":{\\"title\\":\\"My Other App\\",\\"description\\":\\"This is a very simple application.\\",\\"additionalInformation\\":\\"SAPUI5.Component=test.fe.v2.other\\",\\"applicationType\\":\\"URL\\",\\"url\\":\\"/apps/other-app\\",\\"applicationDependencies\\":{\\"manifest\\":true}}}
        };
    </script>

    <script src=\\"../resources/sap/ushell/bootstrap/sandbox2.js\\" id=\\"sap-ushell-bootstrap\\"></script>
    <!-- Bootstrap the UI5 core library. 'data-sap-ui-frameOptions=\\"allow\\"'' is a NON-SECURE setting for test environments -->
    <script id=\\"sap-ui-bootstrap\\"
        src=\\"../resources/sap-ui-core.js\\"
        data-sap-ui-libs=\\"sap.m,sap.ui.core,sap.ushell\\"
        data-sap-ui-async=\\"true\\"
        data-sap-ui-preload=\\"async\\"
        data-sap-ui-theme=\\"sap_horizon\\"
        data-sap-ui-compatVersion=\\"edge\\"
        data-sap-ui-language=\\"en\\"
        data-sap-ui-bindingSyntax=\\"complex\\"
        data-sap-ui-flexibilityServices='[{\\"connector\\":\\"LrepConnector\\",\\"layers\\":[],\\"url\\":\\"/sap/bc/lrep\\"},{\\"applyConnector\\":\\"custom.connectors.WorkspaceConnector\\",\\"writeConnector\\":\\"custom.connectors.WorkspaceConnector\\",\\"custom\\":true},{\\"connector\\":\\"LocalStorageConnector\\",\\"layers\\":[\\"CUSTOMER\\",\\"USER\\"]}]'
        data-sap-ui-resourceroots='{\\"open.ux.preview.client\\":\\"/preview/client\\",\\"test.fe.v2.app\\":\\"/apps/simple-app\\",\\"test.fe.v2.other\\":\\"/apps/other-app\\"}'
        data-sap-ui-frameOptions=\\"allow\\"
        data-sap-ui-xx-componentPreload=\\"off\\"
        data-sap-ui-oninit=\\"module:open/ux/preview/client/flp/init\\">
    </script>

    <script type=\\"text/javascript\\">
        (() => sap.ui.require(['open/ux/preview/client/flp/initConnectors']))()
    </script>

    
</head>

<!-- UI Content -->

<body class=\\"sapUiBody\\" id=\\"content\\">
</body>

</html>",
    "state": "modified",
  },
}
`;

exports[`config generatePreviewFiles tests included and a custom path 1`] = `
Object {
  "webapp/test/flpSandbox.html": Object {
    "contents": "<!DOCTYPE HTML>
<html lang=\\"en\\">
<!-- Copyright (c) 2015 SAP AG, All Rights Reserved -->

<head>
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <title>Local FLP Sandbox</title>

    <!-- Bootstrap the unified shell in sandbox mode for standalone usage.

         The renderer is specified in the global Unified Shell configuration object \\"sap-ushell-config\\".

         The fiori2 renderer will render the shell header allowing, for instance,
         testing of additional application setting buttons.

         The navigation target resolution service is configured in a way that the empty URL hash is
         resolved to our own application.

         This example uses relative path references for the SAPUI5 resources and test-resources;
         it might be necessary to adapt them depending on the target runtime platform.
         The sandbox platform is restricted to development or demo use cases and must NOT be used
         for productive scenarios.
    -->
    <script type=\\"text/javascript\\">
        window[\\"sap-ushell-config\\"] = {
            defaultRenderer: \\"fiori2\\",
            renderers: {
                fiori2: {
                    componentData: {
                        config: {
                            search: \\"hidden\\",
                            enableSearch: false
                        }
                    }
              }
          },
          applications: {\\"myapp-myaction\\":{\\"title\\":\\"My Simple App\\",\\"description\\":\\"This is a very simple application.\\",\\"additionalInformation\\":\\"SAPUI5.Component=test.fe.v2.app\\",\\"applicationType\\":\\"URL\\",\\"url\\":\\"..\\",\\"applicationDependencies\\":{\\"manifest\\":true}}}
        };
    </script>

    <script src=\\"../test-resources/sap/ushell/bootstrap/sandbox.js\\" id=\\"sap-ushell-bootstrap\\"></script>
    <!-- Bootstrap the UI5 core library. 'data-sap-ui-frameOptions=\\"allow\\"'' is a NON-SECURE setting for test environments -->
    <script id=\\"sap-ui-bootstrap\\"
        src=\\"../resources/sap-ui-core.js\\"
        data-sap-ui-libs=\\"sap.m,sap.ui.core,sap.ushell,sap.f,sap.ui.comp,sap.ui.generic.app,sap.suite.ui.generic.template\\"
        data-sap-ui-async=\\"true\\"
        data-sap-ui-preload=\\"async\\"
        data-sap-ui-theme=\\"sap_horizon_dark\\"
        data-sap-ui-compatVersion=\\"edge\\"
        data-sap-ui-language=\\"en\\"
        data-sap-ui-bindingSyntax=\\"complex\\"
        data-sap-ui-flexibilityServices='[{\\"connector\\":\\"LrepConnector\\",\\"layers\\":[],\\"url\\":\\"/sap/bc/lrep\\"},{\\"applyConnector\\":\\"custom.connectors.WorkspaceConnector\\",\\"writeConnector\\":\\"custom.connectors.WorkspaceConnector\\",\\"custom\\":true},{\\"connector\\":\\"LocalStorageConnector\\",\\"layers\\":[\\"CUSTOMER\\",\\"USER\\"]}]'
        data-sap-ui-resourceroots='{\\"open.ux.preview.client\\":\\"/preview/client\\",\\"test.fe.v2.app\\":\\"..\\"}'
        data-sap-ui-frameOptions=\\"allow\\"
        data-sap-ui-xx-componentPreload=\\"off\\"
        data-sap-ui-oninit=\\"module:open/ux/preview/client/flp/init\\">
    </script>

    <script type=\\"text/javascript\\">
        (() => sap.ui.require(['open/ux/preview/client/flp/initConnectors']))()
    </script>

    
</head>

<!-- UI Content -->

<body class=\\"sapUiBody\\" id=\\"content\\">
</body>

</html>",
    "state": "modified",
  },
  "webapp/test/opaTests.qunit.html": Object {
    "contents": "<!DOCTYPE html>
<html>
<head>
	<title>OPA5 tests</title>
	<meta charset=\\"utf-8\\">

	<script
		src=\\"../resources/sap-ui-core.js\\"
		data-sap-ui-theme='sap_horizon_dark'
		data-sap-ui-resourceroots='{
			\\"test.fe.v2.app\\": \\"..\\"
		}'
		data-sap-ui-animation=\\"false\\"
		data-sap-ui-compatVersion=\\"edge\\"
		data-sap-ui-async=\\"true\\">
	</script>

	<script src=\\"opaTests.qunit.js\\"></script>

	<link rel=\\"stylesheet\\" type=\\"text/css\\" href=\\"../resources/sap/ui/thirdparty/qunit-2.css\\">
</head>
<body>
	<div id=\\"qunit\\"></div>
	<div id=\\"qunit-fixture\\"></div>
</body>
</html>",
    "state": "modified",
  },
  "webapp/test/testsuite.qunit.html": Object {
    "contents": "<!DOCTYPE html>
<html lang=\\"en\\">
<head>
    <title>QUnit TestSuite</title>
    <script src=\\"../resources/sap/ui/qunit/qunit-redirect.js\\"></script>
    <script src=\\"/test/testsuite.qunit.js\\" data-sap-ui-testsuite></script>
</head>
<body>
</body>
</html>",
    "state": "modified",
  },
}
`;
